examples/tv-app/tv-common/clusters/messages/MessagesManager.cpp

```diff
@@ -18,39 +18,152 @@
+                memcpy(labelBuffer, response.label.Value().data(), response.label.Value().size());
+
+                ChipLogProgress(Controller, "HandlePresentMessagesRequest option label:%s size:%lu", labelBuffer,
+                                response.label.Value().size());
+
+                optionArray[counter].label = Optional<CharSpan>(CharSpan::fromCharString(labelBuffer));
+            }
+            counter++;
+        }
+
+        Message message{ ByteSpan(messageIdBuffer, messageId.size()),
+                         priority,
+                         messageControl,
+                         startTime,
+                         duration,
+                         CharSpan::fromCharString(messageTextBuffer),
+                         Optional<DataModel::List<MessageResponseOption>>(
+                             DataModel::List<MessageResponseOption>(optionArray, size)) };
+
+        mMessages.push_back(message);
+    }
+    else
+    {
+        Message message{ ByteSpan(messageIdBuffer, messageId.size()), priority, messageControl, startTime, duration,
+                         CharSpan::fromCharString(messageTextBuffer) };
+        mMessages.push_back(message);
+    }
+
     // Add your code to present Message
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest complete");
+
+    return;
+exit:
+
+    ChipLogProgress(Controller, "HandlePresentMessagesRequest error exit");
+    // only reach here on errors
+    if (messageIdBuffer != nullptr)
+    {
+        delete[] messageIdBuffer;
+    }
+    for (MessageResponseOption * memloc : optionToFree)
+    {
+        delete[] memloc;
+    }
+    for (char * memloc : needToFree)
+    {
+        delete[] memloc;
+    }
 }
 
-void MessagesManager::HandleCancelMessagesRequest(const chip::app::DataModel::DecodableList<chip::ByteSpan> & messageIds)
+void MessagesManager::HandleCancelMessagesRequest(const DataModel::DecodableList<ByteSpan> & messageIds)
 {
-    // TODO: Cancel Message
+    auto iter = messageIds.begin();
+    while (iter.Next())
+    {
+        auto & id = iter.GetValue();
+
+        mMessages.remove_if([id](Clusters::Messages::Structs::MessageStruct::Type & entry) {
+            if (entry.messageID.data_equal(id))
+            {
+                // free the memory allocated for this entry
+                if (entry.responses.HasValue())
+                {
+                    for (const MessageResponseOption & option : entry.responses.Value())
+                    {
+                        if (option.label.HasValue())
+                        {
+                            delete[] option.label.Value().data();
+                        }
+                    }
+                    delete[] entry.responses.Value().data();
+                }
+                delete[] entry.messageID.data();
+                delete[] entry.messageText.data();
+                return true;
+            }
+            return false;
+        });
+    }
```

Comment:
Where's the check for iterator success?

Reply:
added comment about how a match does not need to be found to return success